service/storage_proxy.cc

```diff
@@ -446,51 +447,67 @@ class storage_proxy::remote {
-                    // send buffer.
-                    auto f = co_await coroutine::as_future(send_mutation_done(netw::messaging_service::msg_addr{reply_to, shard}, trace_state_ptr,
-                            shard, response_id, p->get_view_update_backlog()));
-                    f.ignore_ready_future();
-                } catch (...) {
-                    std::exception_ptr eptr = std::current_exception();
-                    errors.count++;
-                    errors.local = replica::try_encode_replica_exception(eptr);
-                    seastar::log_level l = seastar::log_level::warn;
-                    if (is_timeout_exception(eptr) || std::holds_alternative<replica::rate_limit_exception>(errors.local.reason)) {
-                        // ignore timeouts and rate limit exceptions so that logs are not flooded.
-                        // database's total_writes_timedout or total_writes_rate_limited counter was incremented.
-                        l = seastar::log_level::debug;
-                    }
-                    slogger.log(l, "Failed to apply mutation from {}#{}: {}", reply_to, shard, eptr);
-                }
-            },
-            [&] {
-                // Note: not a coroutine, since often nothing needs to be forwarded and this returns a ready future
-                return parallel_for_each(forward.begin(), forward.end(), [&] (gms::inet_address forward) {
-                    // Note: not a coroutine, since forward_fn() typically returns a ready future
-                    tracing::trace(trace_state_ptr, "Forwarding a mutation to /{}", forward);
-                    return forward_fn(p, netw::messaging_service::msg_addr{forward, 0}, timeout, m, reply_to, shard, response_id,
-                                        tracing::make_trace_info(trace_state_ptr))
-                            .then_wrapped([&] (future<> f) {
-                        if (f.failed()) {
-                            ++p->get_stats().forwarding_errors;
-                            errors.count++;
-                        };
+
+        if (auto stale = try_apply_fence(fence); stale) {
+            errors.count += (forward.size() + 1);
+            errors.local = std::move(*stale);
+        }
+
+        if (errors.count == 0) {
+            co_await coroutine::all(
+                [&]() -> future<> {
+                    try {
+                        // FIXME: get_schema_for_write() doesn't timeout
+                        schema_ptr s = co_await get_schema_for_write(schema_version, netw::messaging_service::msg_addr{reply_to, shard}, timeout);
+                        // Note: blocks due to execution_stage in replica::database::apply()
+                        co_await apply_fn(p, trace_state_ptr, std::move(s), m, timeout);
+
+                        if (auto stale = try_apply_fence(fence); stale) {
+                            throw std::move(*stale);
```

Comment:
Why is this repeated?

Reply:
Ah, you say why. I think it deserves a comment here.

Reply:
The idea is to send failure to the request coordinator so it knows that it may have written using stale topology. Not sure this is needed if we drain without a timeout.

Reply:
But why try_apply_fence twice?

Reply:
> Not sure this is needed if we drain without a timeout.

Currently it's not needed since we `barrier_and_drain` all replicas. It's included to resolve @kbr-scylla case from the link above, if in the future we would drain only some affected replicas.

Reply:
> But why try_apply_fence twice?

We can fence only once after the op, but then you may do unnecessary work. It is like optimistic locking.

Reply:
How can you fence after the op, it if's a write? Isn't it to late?

Since we grab the reference to the topology, isn't it enough to check it at the beginning? The fence rpc will wait until we release it.

Reply:
You need either wait in the fence rpc for old version to complete or re-check after the op and drop a reply if fence changed in between. Here second option is implemented. Is more live (you do not need to wait while changing topology) and waiting for replica writes will be needed only before running cleanup.